source/common/config/filesystem_subscription_impl.h

```diff
@@ -43,28 +48,42 @@ class FilesystemSubscriptionImpl : public Config::Subscription<ResourceType> {
   }
 
 private:
-  void refresh() {
+  void refresh(bool ignore_read_failure) {
+    ENVOY_LOG(debug, "Filesystem config refresh for {}", path_);
+    stats_.update_attempt_.inc();
     try {
       const std::string json = Filesystem::fileReadToEnd(path_);
       envoy::api::v2::DiscoveryResponse message;
       const auto status = google::protobuf::util::JsonStringToMessage(json, &message);
       if (status != google::protobuf::util::Status::OK) {
-        // TODO(htuch): Track stats and log failures.
         callbacks_->onConfigUpdateFailed(nullptr);
+        ENVOY_LOG(warn, "Filesystem config JSON conversion error: {}", status.ToString());
+        stats_.update_failure_.inc();
         return;
       }
       const auto typed_resources = Config::Utility::getTypedResources<ResourceType>(message);
-      callbacks_->onConfigUpdate(typed_resources);
-      // TODO(htuch): Add some notion of current version for every API in stats/admin.
+      try {
+        callbacks_->onConfigUpdate(typed_resources);
+        stats_.update_success_.inc();
+        // TODO(htuch): Add some notion of current version for every API in stats/admin.
+      } catch (const EnvoyException& e) {
```

Comment:
The above comment was really aimed at this. I was thinking it could be useful to have more than one exception type, but I think there's probably a fair bit of logic buried in config processing today around exceptions that expect to just be able to throw `EnvoyException`.